apis/bigqueryreservation/v1alpha1/assignment_identity.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/apis/common"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AssignmentIdentity defines the resource identity to BigQueryReservationAssignment,
+// which is to be used in the controller.
+type AssignmentIdentity struct {
+	// The reservation to which the assignment is currently attached
+	parent *BQReservation
+	// The assignment resourceID
+	id string
+}
+
+func (i *AssignmentIdentity) String() string {
+	return i.parent.String() + "/assignments/" + i.id
+}
+
+func (i *AssignmentIdentity) ID() string {
+	return i.id
+}
+
+func (i *AssignmentIdentity) Parent() *BQReservation {
+	return i.parent
+}
+
+func (i *AssignmentIdentity) GetProjetID() string {
+	return i.parent.ProjectID
+}
+
+type BQReservation struct {
+	ProjectID       string
+	Location        string
+	ReservationName string
+}
+
+func (p *BQReservation) String() string {
+	return "projects/" + p.ProjectID + "/locations/" + p.Location + "/reservations/" + p.ReservationName
+}
+
+// New builds a AssignmentIdentity from the Config Connector BigQueryReservationAssignment object.
+func NewAssignmentIdentity(ctx context.Context, reader client.Reader, obj *BigQueryReservationAssignment) (*AssignmentIdentity, error) {
+
+	// Get the reservation to move the assignment to.
+	// Uses the "spec.ReservationRef"
+	var name string
+	var err error
+	if obj.Spec.ReservationRef.External != "" {
+		name = obj.Spec.ReservationRef.External
+	} else {
+		name, err = obj.Spec.ReservationRef.NormalizedExternal(ctx, reader, obj.GetNamespace())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	reservationParent, reservationID, err := ParseReservationExternal(name)
+	if err != nil {
+		return nil, err
+	}
+
+	parent := &BQReservation{
+		ProjectID:       reservationParent.ProjectID,
+		Location:        reservationParent.Location,
+		ReservationName: reservationID,
+	}
+
+	// Get desired assignment resourceID
+	// Only works for resource acquisition
+	resourceID := common.ValueOf(obj.Spec.ResourceID)
```

Comment:
if `obj.Spec.ResourceID` is set (acquisition), use it as `resourceID`. Otherwise we should check if `status.externalRef` is populated. If it is populated (meaning the resource was already created, and we are in the subsequent reconciliations), use the ID parsed from `status.externalRef`.